File: contest/LeetCode_Weekly_Contest_201/Minimum_Cost_to_Cut_a_Stick.cpp
class Solution {
public:
    int dp[101][101];
    int n;
    int dfs(int st, int ed, const vector<int>& cuts) {
        if (ed < st) return 0;

        if (dp[st][ed] != -1) return dp[st][ed];

        int L = -1, R = -1;
        if (st == 0) L = 0;
        else L = cuts[st-1];

        if (ed == cuts.size()-1) R = n;
        else R = cuts[ed+1];

        int& ret = dp[st][ed];
        ret = INT_MAX;
        int val = R - L;
        for (int i = st; i <= ed; ++i) if (L < cuts[i] && cuts[i] < R) {
            ret = min(ret, dfs(st, i-1, cuts) + dfs(i+1, ed, cuts) + val);
        }

        if (ret == INT_MAX) ret = 0;
        return ret;
    }
    int minCost(int n, vector<int>& cuts) {
        sort(cuts.begin(), cuts.end());
        this->n = n;
        memset(dp, -1, sizeof dp);
        return dfs(0, cuts.size()-1, cuts);
    }
};

File: contest/LeetCode_Weekly_Contest_201/Minimum_Cost_to_Cut_a_Stick.py
class Solution(object):
    def dfs(self, st, ed):
        if ed < st: return 0
        
        if self.dp[st][ed] != -1:
            return self.dp[st][ed]
        
        L, R = 0, self.n
        if st != 0: 
            L = self.cuts[st-1]
        if ed != self.m-1: 
            R = self.cuts[ed+1]
        
        val, ret = R - L, 1<<32
        
        for i in xrange(st, ed+1):
            if L < self.cuts[i] and self.cuts[i] < R:
                ret = min(ret, self.dfs(st, i-1) + self.dfs(i+1, ed) + val)
        
        if ret == 1<<32:
            ret = 0
        self.dp[st][ed] = ret
        return ret
        
    def minCost(self, n, cuts):
        """
        :type n: int
        :type cuts: List[int]
        :rtype: int
        """
        cuts.sort()
        self.n = n
        self.m = len(cuts)
        self.cuts = cuts
        self.dp = [[-1 for _ in range(self.m)] for _ in xrange(self.m)]
        
        return self.dfs(0, self.m-1)
        
